core/src/commonMain/kotlin/com/powersync/attachments/sync/SyncingService.kt

```diff
@@ -0,0 +1,311 @@
+package com.powersync.attachments.sync
+
+import co.touchlab.kermit.Logger
+import com.powersync.PowerSyncException
+import com.powersync.attachments.Attachment
+import com.powersync.attachments.AttachmentContext
+import com.powersync.attachments.AttachmentService
+import com.powersync.attachments.AttachmentState
+import com.powersync.attachments.LocalStorage
+import com.powersync.attachments.RemoteStorage
+import com.powersync.attachments.SyncErrorHandler
+import com.powersync.utils.throttle
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancelAndJoin
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.flow.buffer
+import kotlinx.coroutines.flow.merge
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
+
+/**
+ * Service responsible for syncing attachments between local and remote storage.
+ *
+ * This service handles downloading, uploading, and deleting attachments, as well as
+ * periodically syncing attachment states. It ensures proper lifecycle management
+ * of sync operations and provides mechanisms for error handling and retries.
+ *
+ * @property remoteStorage The remote storage implementation for handling file operations.
+ * @property localStorage The local storage implementation for managing files locally.
+ * @property attachmentsService The service for managing attachment states and operations.
+ * @property getLocalUri A function to resolve the local URI for a given filename.
+ * @property errorHandler Optional error handler for managing sync-related errors.
+ * @property logger Logger instance for logging sync operations and errors.
+ * @property syncThrottle The minimum duration between consecutive sync operations.
+ * @property scope The coroutine scope used for managing sync operations.
+ */
+public class SyncingService
+    @OptIn(DelicateCoroutinesApi::class)
```

Comment:
IntelliJ says we don't need this annotation here :)

Reply:
This was left from experimenting with defaulting the coroutine context to `GlobalScope`. Good catch, I've removed this now.